### PR TITLE
New velocity export

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 1]
```diff
@@ -5,14 +5,23 @@
...
 
 import requests
 
-
 PROGRAM_CODE = 'FS'  # Our Full-Stack program
+#BREADCRUMB_INDEX_URL = settings.LMS_SYLLABUS
+BREADCRUMB_INDEX_URL = settings.BREADCRUMB_INDEX_URL
+KEYS = ['module','section','lesson']
+
+PROJECTS = {
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Can you please add a TODO comment here that we should look into creating a way to update these dynamically in a central location (perhaps as part of the the new weightings API)? - I wouldn't want us to have to maintain these fractions in multiple places

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 2]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
             for module, timestamps in per_module_lessons_times.items()}
 
 
+def fourteen_days_fractions(completed_fractions):
+    fourteen_days_ago = timezone.now() - timedelta(days=14)
+    return sum(item['lesson_fraction'] if item['time_completed'] > fourteen_days_ago else 0
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Line too long - please aim to fit everything within the PEP8 80 character limit.

Let me tell you how my mind works when looking at these things:
- hmm... the line is too long, what's taking up so much space?
- well, it's a single ternary expression, no nice way to split it directly
- anything else?
- well, we can add a break after the `sum(` bracket - ah, no, it's still too long
- what is actually taking so much space - ah, it's the variable `fourteen_days_ago`
- why does it have to be so long?
- actually, why is it so specific?
- it's baked into the function name - is there a good reason that the function name is so specific?
- well, it might have been in the past, but now that we're looking to add 28 days as well, maybe it doesn't need to be that specific - how about perhaps `def n_days_fractions(completed_fractions, days_ago=14)`
- if we go with that, then that both makes the code:
    - more future proof (though this is not always a good thing - we need to strike a balance between extensibility and complexity)
    - and being shorter to read - with the potential to change the variable to something like `start_time` or `earliest_time` or `period_start`
- any of the above then, together with the line break after the bracket would make it fit, e.g.:
```
def n_days_fractions(completed_fractions, days_ago=14):
    period_start = timezone.now() - timedelta(days=days_ago)
    return sum(
        item['lesson_fraction'] if item['time_completed'] > period_start else 0
        for item in completed_fractions)
```
In other words, I believe that though often annoying, this 80 character limitation is very effective in forcing us to review our understanding and ability to reason about the relevant piece of code.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 3]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
 
 
+def fourteen_days_fractions(completed_fractions):
+    fourteen_days_ago = timezone.now() - timedelta(days=14)
+    return sum(item['lesson_fraction'] if item['time_completed'] > fourteen_days_ago else 0 
+                for item in completed_fractions)
+
+
+def cumulative_days_fractions(completed_fractions):
+    return sum(item['lesson_fraction'] for item in completed_fractions)
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: This function again makes me think about the above - given that they're very similar, maybe we should combine them? Perhaps with this one being a special case of the above, with the `days_ago` omitted (changing the default from 14 to None)?
Up to you.

Also, please note that I'm aware that much of this is copied from AMOS - it's just that even if we deemed the code to be good in the past, every review should be an opportunity to tidy up. ("Leave it better than you found it")

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 4]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+
+
+def fractions_per_day(date_joined, limit, completed_fractions):
+        """Creates an array of fractions completed for each day since the student started
+        
+        Returns a single string of all the array elements joined by a comma
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: This line kinda contradicts the above.
The first line of the docstring should summarise the entire function, and the other lines should provide additional details - so it's confusing that the above says it's an array and now it says it's a string.

I don't have a particular suggestion here, just a request that you try to make it more consistent.

Two other comments:
- The convention that I'd like to follow for the first line is of an imperative sentence - i.e "Create", rather than "Creates" - that's what's generally more common in Python, and I think it's more fun to tell the function what to do :)
- Instead of "a single string of x's joined by a comma" you can say "comma separated x's"

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 5]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+        """
+        range_limit = (timezone.now() - date_joined).days
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: It's generally preferred to name variables after what they represent rather than what they'll be used for. I'd suggest something like `days_since_joined` here :)

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 6]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+        # Needs to be string, then cast to float for calculation
+        # Then be converted back to string for join operation
+        fractions_days = {str(i) : '0' for i in range(range_limit + 1)}
+        for item in completed_fractions:
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: I would suggest avoiding general indexes like `i` or `item` when the function deals with more specific elements. In this case, I would probably go with `fraction` or `f` (if I want it to be short) - I find that it helps reason about what the function does.

But looking at the rest of the function - this actually wouldn't work, since then you'd have `fraction[lesson_fraction]` - so this forces us to think about what kind of thing it is that is stored in `completed_fractions`.

The easy way out would be to call it something like `fraction_record`, but if possible, it's best to come up with a single word name that would describe it well. Alternatively, a good way to think about it is - if you were writing in something like Java or C++ and you had to create a class for this - what would be the class's name?

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 7]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+            days_in = str((item['time_completed'] - date_joined).days)
+            fractions_days[days_in] = str(float(fractions_days[days_in])
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: You know what - maybe I'm a bit tired, but I feel that even after looking at this function for a while, I still don't quite have a good understanding of how it works. If you can describe the algorithmic approach relatively simply, please add it to the docstring. If not, I'd suggest refactoring.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 8]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+
+
+def completed_fraction_per_module(fractions, completed_fractions):
+    """Aggregates completed fractions witin last 14d and before that
+        
+    Returns a dict with module and the completed aggregations
+    """
+    fourteen_days_ago = timezone.now() - timedelta(days=14)
+    for key, item in completed_fractions.items():
+        accessor = format_module_field(key[0], '_fraction_within_14d')
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Lines too long - please see what can be done with this

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 9]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+        fractions_days = {str(i) : '0' for i in range(range_limit + 1)}
+        for item in completed_fractions:
+            days_in = str((item['time_completed'] - date_joined).days)
+            fractions_days[days_in] = str(float(fractions_days[days_in]) 
+                                            + item['lesson_fraction'])
+        return ','.join(OrderedDict(sorted(fractions_days.items())).values())
+
+
+def completed_fraction_per_module(fractions, completed_fractions):
+    """Aggregates completed fractions witin last 14d and before that
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: There's a typo there in the docstring with `witin`
Also, I have no idea what it means - `within last 14d and before that` sounds like an oxymoron. Did you mean to say something like `Separate completed fractions into those in last 14d and the rest`?

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 8 above)
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: What does the term "accessor" mean in this context?

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 10]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+                                        if item['time_completed'] > fourteen_days_ago 
+                                        else format_module_field(key[0], 
+                                        '_fraction_before_14d')
+        if accessor in fractions:
+            fractions[accessor] += item['lesson_fraction']
+    return fractions
+
+
+def create_fractions_dict(syllabus):
+    fractions = {format_module_field(x['module'],'_fraction_within_14d') : 0 for x in syllabus.values()}
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: - Missing docstring.
- Undescriptive variable name `x`
    - And the variable `syllabus` doesn't help either - what is a syllabus a collection of? lessons?

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 11]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+    fractions.update({format_module_field(x['module'],'_fraction_before_14d') : 0 for x in syllabus.values()})
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: too long

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 12]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+    return fractions
+
+
+def completed_percent_per_module(suffix, fractions, module_fractions):
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: missing docstring

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 13]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+    for module, module_fraction in module_fractions.items():
+        accessor = format_module_field(module, suffix)
+        if accessor in fractions and module_fraction != 0:
+            fractions[accessor] = fractions[accessor] / (module_fraction + (PROJECTS[module]
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: the line here is too long and I don't quite understand the logic.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 14]
```diff
@@ -86,12 +95,72 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+                                                            if module in PROJECTS else 0.0))
+    return fractions
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Also, if we use the term `fractions` everywhere, it's probably best to use that in the function name too, rather than `percent`

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 15]
```diff
@@ -117,6 +189,24 @@ def all_student_data(program):
                 # for each lesson learned, store latest timestamp
                 completed_lessons[breadcrumbs] = activity.modified
 
+                #Calculate fractions
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: When an inner part of a function (e.g. inside a loop) requires a header comment, that's generally a sign that it should be extracted into a function - and I think this is one of these cases

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 16]
```diff
@@ -5,14 +5,23 @@
...
-from collections import Counter, defaultdict
+from collections import Counter, defaultdict, OrderedDict
 from datetime import datetime, timedelta
 import json
 
 import requests
 
-
 PROGRAM_CODE = 'FS'  # Our Full-Stack program
+#BREADCRUMB_INDEX_URL = settings.LMS_SYLLABUS
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Why is this commented out? If it's no longer relevant, then let's remove it.

**Review by yoniLavi — CHANGES_REQUESTED** [review]

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 1 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Added a comment to remind us to get the project scores dynamically.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 2 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: I agree with keeping it down to 80 characters and your logic around naming and function reuse. I changed the function to accommodate for n days and the cumulative.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 3 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Removed this function as it is now incorporated in n_days_fractions. Made `days_ago` 0 by default and added `or days_ago == 0` to the conditional. `None` would have either broken the code or required an extra condition and logically we should not be looking for fractions for 0 days ago. Therefore I used 0, but I am open to changing this again.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 16 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Removed this.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 4 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Changed the docstring. Please let me know if it is better now, I am not great with these. Maybe we can do a quick 10 mins session on these and add something to our documentation.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 5 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Changed the variable name as suggested.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 6 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: okay, well maybe I should even rename this to simply `days` as this creating a dict with an entry per day in the student's lifecycle.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 7 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: I added a step-by-step explanation that will hopefully help understand the code better. I also renamed some of the variables. Please let me know if this makes sense now. If not, let's sync if there is a better way to do this.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 8 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Split this into multiple lines

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 9 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Updated the docstring as suggests and yes I meant separate completed fractions into in last 14d and the rest.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 8 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Could not call it key because there would have been a conflict. Renamed key, item to make more sense and changed accessor to key. also changed the function to take the `days_ago` and dynamically generate fields and change `fourteen_days_ago` to `n_days_ago`

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 10 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Added docstring. Yes, syllabus is the information that is retrieved from the LMS Syllabus Google sheet API that has the fractions and the module - section - lesson association. Changed syllabus to lessons and x to lesson to make it more readable.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 11 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Split the code into multiple lines.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 12 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Added a docstring to the function.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 13 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Split into multiple lines and added the logic to the docstring for further clarification. Let me know it makes sense or you want more information about this.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 14 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: This is to specifically distinguish between the two fraction is fraction of the overall whereas percent I mean percent of the module completed. I know that they technically are both fractions or both percent, but I want to make the distinction here. Let me know if you want me to still change it.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 15 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Created new function for this inner part of the function.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 14 above)
Comment (yoniLavi) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Interesting distinction. I'm very happy to keep it, as long as we're careful to use it consistently.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 17]
```diff
@@ -86,12 +95,132 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+        cumulative_fraction = lesson['fractions']['cumulative_fraction']
+    
+    completed_fractions[breadcrumbs] = {
+        'time_completed' : modified_time,
+        'lesson_fraction' : lesson_fraction,
+        'module_fraction' : module_fraction,
+        'cumulative_fraction' : cumulative_fraction}
+
+
+
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Extra blank line :)

**Review by yoniLavi — DISMISSED** [review]
This is excellent, thanks @stefan-ci !

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 17 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Removed the extra blank line :)

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py (hunk 14 above)
Comment (ghost) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Actually, I had to remove the percentage calculation. This means we will need to calculate the percentage of the module in AMOS, but otherwise we will not know the actual fraction breakdown when adding the project fraction.

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 18]
```diff
@@ -86,12 +87,110 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+def n_days_fractions(completed_fractions, days_ago=0):
+    """Sum fractions completed for the n previous days
+    If days_ago is 0 (by default) it will sum all completed fractions
+
+    Returns the sum of fractions as float
+    """
+    period_start = timezone.now() - timedelta(days=days_ago)
+    return sum(
+        item['lesson_fraction'] 
+        if item['time_completed'] > period_start or days_ago == 0 else 0
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Suggest refactoring this expression, to give the predicate a name, either as a function that is called by the comprehension, or as a variable in a loop

Comment (yoniLavi) [reply lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Maybe even better, I'd suggest changing the sentinel value to `None`, and having the condition occur before the summation, such that for `None` the `period_start` would be the beginning of time.

**Review by yoniLavi — DISMISSED** [review]

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 19]
```diff
@@ -87,17 +87,19 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
 
-def n_days_fractions(completed_fractions, days_ago=0):
+def n_days_fractions(completed_fractions, days_ago=None):
     """Sum fractions completed for the n previous days
-    If days_ago is 0 (by default) it will sum all completed fractions
 
     Returns the sum of fractions as float
     """
-    period_start = timezone.now() - timedelta(days=days_ago)
+    if days_ago == None:
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py]: Hi @stefan-ci , sorry about nitpicking, but it's considered bad practice to do `== None`. Instead, since `None` is a special singleton, we should use `is None`.
Would you mind please changing that?

File: lms/djangoapps/learning_success/management/commands/export_all_activity_records.py [hunk 20]
```diff
@@ -87,17 +87,19 @@ def lessons_days_into_per_module(first_active, breadcrumb_dict):
...
+        period_start = datetime.utcfromtimestamp(0)
+    else:
+        period_start = timezone.now() - timedelta(days=days_ago)
     return sum(
-        item['lesson_fraction'] 
-        if item['time_completed'] > period_start or days_ago == 0 else 0 
-        for item in completed_fractions)
+        item['lesson_fraction']
+        for item in completed_fractions
+        if item['time_completed'] > period_start)
```
Comment (yoniLavi) [inline lms/djangoapps/learning_success/management/commands/export_all_activity_records.py:102]: This is much nicer, thanks!

**Review by yoniLavi — APPROVED** [review]
